File: solutions/clojure/014-3sum-closest.clj
;; Problem 14: 3Sum Closest (LeetCode 16)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn three-sum-closest [nums target]
  (let [sorted (vec (sort nums))
        n (count sorted)]
    (reduce
      (fn [closest i]
        (loop [l (inc i), r (dec n), best closest]
          (if (>= l r) best
            (let [sum (+ (sorted i) (sorted l) (sorted r))
                  better (if (< (Math/abs (- sum target))
                               (Math/abs (- best target)))
                           sum best)]
              (cond
                (< sum target) (recur (inc l) r better)
                (> sum target) (recur l (dec r) better)
                :else sum)))))
      (+ (sorted 0) (sorted 1) (sorted 2))
      (range (- n 2)))))

File: solutions/scala/014-3sum-closest.scala
/**
 * Problem 14: 3Sum Closest (LeetCode 16)
 * Difficulty: Med
 * Language: Scala
 */
def threeSumClosest(nums: Array[Int], target: Int): Int = {
  val sorted = nums.sorted
  val n = sorted.length

  @annotation.tailrec
  def search(i: Int, l: Int, r: Int, closest: Int): Int = {
    if (l >= r) closest
    else {
      val sum = sorted(i) + sorted(l) + sorted(r)
      val better = if (math.abs(sum - target) < math.abs(closest - target)) sum else closest
      if (sum < target) search(i, l + 1, r, better)
      else if (sum > target) search(i, l, r - 1, better)
      else sum
    }
  }

  (0 until n - 2).foldLeft(Int.MaxValue) { (closest, i) =>
    search(i, i + 1, n - 1, closest)
  }
}
